zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -286,6 +293,91 @@ void rollbackPendingChanges(long zxid, Map<String, ChangeRecord>pendingChangeRec
         }
     }
 
+    /**
+     * check whether exceeded the quota.
+     *
+     * @param lastPrefix
+     *            the path of the node that is quotaed.
+     * @param bytesDiff
+     *            the diff to be added to number of bytes
+     * @param countDiff
+     *            the diff to be added to the count
+     */
+    private void checkQuota(String lastPrefix, long bytesDiff, int countDiff)
+            throws KeeperException.QuotaExceededException {
+        if (!enforeQuota) {
+            return;
+        }
+
+        LOG.debug("checkQuota: lastPrefix={}, bytesDiff={}, countDiff={}", lastPrefix, bytesDiff, countDiff);
+
+        if (lastPrefix == null || lastPrefix.length() == 0) {
+            return;
+        }
+        // now check if the counts match the quota
+        String limitNode = Quotas.limitPath(lastPrefix);
+        DataNode node = zks.getZKDatabase().getNode(limitNode);
+        StatsTrack limitStats;
+        if (node == null) {
+            // should not happen
+            LOG.error("Missing count node for quota " + limitNode);
+            return;
+        }
+        synchronized (node) {
+            limitStats = new StatsTrack(new String(node.data));
```

Comment:
This does the parsing & "newing" with the node locked; I'd just get the data and move the constructors out of the `synchronized` block.

Reply:
If I remember correctly, if don't `synchronized (node)`, `spotbugs` will complain about it